src/renderer/html_handlebars/toc_json.rs

```diff
@@ -0,0 +1,195 @@
+use std::cmp;
+use std::collections::BTreeMap;
+use std::path::Path;
+
+use pulldown_cmark::{html, Parser, Event, Tag};
+use serde_json;
+use errors::*;
+
+fn path_to_link(path: &str) -> String {
+    Path::new(path)
+        .with_extension("html")
+        .to_str()
+        .unwrap()
+        // Hack for windows who tends to use `\` as separator instead of `/`
+        .replace("\\", "/")
+}
+
+fn strip_name(name: &str) -> String {
+    // filter all events that are not inline code blocks
+    let parser = Parser::new(name).filter(|event| match *event {
+        Event::Start(Tag::Code) |
+            Event::End(Tag::Code) |
+            Event::InlineHtml(_) |
+            Event::Text(_) => true,
+        _ => false,
+    });
+
+    // render markdown to html
+    let mut markdown_parsed_name = String::with_capacity(name.len() * 3 / 2);
+    html::push_html(&mut markdown_parsed_name, parser);
+
+    markdown_parsed_name
+}
+
+/// Create a json node for a given chapter.
+fn make_chapter_node(item: &BTreeMap<String, String>) -> serde_json::Value {
+    let mut map = BTreeMap::new();
+
+    if let Some(path) = item.get("path") {
+        if !path.is_empty() {
+            map.insert("link".to_owned(), json!(path_to_link(path)));
+        }
+    }
+
+    if let Some(section) = item.get("section") {
+        map.insert("section".to_owned(), json!(section));
+    }
+
+    if let Some(name) = item.get("name") {
+        map.insert("name".to_owned(), json!(strip_name(name)));
+    }
+
+    if let Some(previous_path) = item.get("previous_path") {
+        map.insert("previous".to_owned(),
+                   json!({
+                             "link": path_to_link(previous_path)
+                         }));
+    }
+
+    if let Some(next_path) = item.get("next_path") {
+        map.insert("next".to_owned(),
+                   json!({
+                             "link": path_to_link(next_path)
+                         }));
+    }
+
+    json!(map)
+}
+
+/// Extend or collapse tree path to reach a certain depth.
+fn extend_or_collapse_path(level: usize, path: &mut Vec<serde_json::Value>) {
+    // Can't pop root node
+    assert!(level > 0);
+
+    while level > path.len() {
+        path.push(json!({}));
+    }
+
+    while level < path.len() {
+        // Push child into parent.children
+        let child = path.pop().unwrap();
+        let parent = path.last_mut().unwrap().as_object_mut().unwrap();
+
+        if !parent.contains_key("children") {
+            parent.insert("children".to_owned(), json!([]));
+        }
+
+        parent.get_mut("children")
+              .unwrap()
+              .as_array_mut()
+              .unwrap()
+              .push(child);
+    }
+}
+
+/// Turn a flat chapters array into a tree json structure. Each node represents
+/// a section and its subsections.
+pub fn make_toc_tree(chapters: &[BTreeMap<String, String>]) -> Result<serde_json::Value> {
```

Comment:
I like the more descriptive name and the doc-comment you've added. Now we can tell at a glance what this function does :)